src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,607 @@
+        {
+            using (var stream = await GetStreamAsync(uri, log, token))
+            using (var reader = new StreamReader(stream))
+            using (var jsonReader = new JsonTextReader(reader))
+            {
+                return JObject.Load(jsonReader);
+            }
+        }
+
+        private async Task<HttpResponseMessage> SendWithCredentialSupportAsync(
+            HttpRequestMessage request,
+            HttpCompletionOption completionOption,
+            CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = null;
+            ICredentials promptCredentials = null;
+
+            // Keep a local copy of the http client, this allows the thread to check if another thread has
+            // already added new credentials.
+            var httpClient = _httpClient;
+
+            // Create the http client on the first call
+            if (httpClient == null)
+            {
+                await HttpClientLock.WaitAsync();
+                try
+                {
+                    // Double check
+                    if (_httpClient == null)
+                    {
+                        _httpClient = await CreateHttpClient();
+                        httpClient = _httpClient;
+                    }
+                }
+                finally
+                {
+                    HttpClientLock.Release();
+                }
+            }
+
+            // Update the request for STS
+            STSAuthHelper.PrepareSTSRequest(_baseUri, CredentialStore.Instance, request);
+
+            // Authorizing may take multiple attempts
+            while (true)
+            {
+                // Clean up any previous responses
+                if (response != null)
+                {
+                    response.Dispose();
+                }
+
+                // Read the response headers before reading the entire stream to avoid timeouts from large packages.
+                response = await httpClient.SendAsync(
+                    request,
+                    HttpCompletionOption.ResponseHeadersRead,
+                    cancellationToken);
+
+                if (response.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    // Create a copy of the request for the next call.
+                    // Requests may only be sent once.
+                    request = CloneRequest(request);
+
+                    try
+                    {
+                        // Only one request may prompt and attempt to auth at a time
+                        await HttpClientLock.WaitAsync();
+
+                        // Auth may have happened on another thread, if so just continue
+                        if (!object.ReferenceEquals(httpClient, _httpClient))
+                        {
+                            continue;
+                        }
+
+                        // Give up after 10 tries.
+                        _authRetries++;
```

Comment:
Isn't this the second place we're doing this? Here and `HttpHandlerResourceV3Provider`.